tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+//! # Implementation Details
+//!
+//! The semaphore is implemented using an intrusive linked list of waiters. An
+//! atomic counter tracks the number of available permits. If the semaphore does
+//! not contain the required number of permits, the task attempting to acquire
+//! permits places its waker at the end of a queue. When new permits are made
+//! available (such as by releasing an initial acquisition), they are assigned
+//! to the task at the front of the queue, waking that task if its requested
+//! number of permits is met.
+//!
+//! Because waiters are enqueued at the back of the linked list and dequeued
+//! from the front, the semaphore is fair. Tasks trying to acquire large numbers
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+use crate::loom::cell::CausalCell;
+use crate::loom::sync::{atomic::AtomicUsize, Mutex, MutexGuard};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::future::Future;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::atomic::Ordering::*;
+use std::task::Poll::*;
+use std::task::{Context, Poll, Waker};
+use std::{cmp, fmt};
+
+/// An asynchronous counting semaphore which permits waiting on multiple permits at once.
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting {
+        /// We are waiting to acquire this many permits.
+        needed: u16,
+        /// We have previously acquired this many permits.
+        acquired: u16,
+    },
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter.
+    ///
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: CausalCell<usize>,
```

Comment:
I did not  consider the additional unsafety required to avoid `AtomicUsize`. My impression is that avoiding the unsafety would be better.. i.e. switch back to an atomic 👼 . However, I won't block on this either way, so it is your choice.

Reply:
That said, it looks like there is some common aspects with `Notify`. We could possibly implement a `WaitList<T>` struct that is able to contain a generic `T` that is guarded by the list mutex.

Also, not a requirement for this PR.

Reply:
>  My impression is that avoiding the unsafety would be better.. i.e. switch back to an atomic

Yeah. Ideally we would just express this safely by making the methods that mutate the state just take mutable references. But, because the wait list entry is accessible via both the future that owns it, _and_ through the list, we can't use mutable/shared refs to express this (we could remove a few uses of the `unsafe` _keyword by not using `CausalCell`s, but the unsafety is still there and we still need to uphold additional invariants). 

I think it's probably better to bring back the atomic (it also avoids acquiring the lock in at least one codepath, and has a small but measurable perf impact in my benchmarks).

Reply:
> That said, it looks like there is some common aspects with `Notify`. We could possibly implement a `WaitList<T>` struct that is able to contain a generic `T` that is guarded by the list mutex.
> 
> Also, not a requirement for this PR.

I like this idea, but I think it's best saved for a follow-up change IMO.

Reply:
That is fine. I'm just writing out my thoughts.